python_modules/dagster/dagster/_core/definitions/cacheable_assets.py

```diff
@@ -104,3 +106,56 @@ def build_definitions(
     ) -> Sequence[AssetsDefinition]:
         """For a given set of AssetsDefinitionMetadata, return a list of AssetsDefinitions"""
         raise NotImplementedError()
+
+    def with_resources(
+        self, resource_defs: Mapping[str, ResourceDefinition]
+    ) -> "CacheableAssetsDefinition":
+        return WrappedCacheableAssetsDefinition(
+            self._unique_id,
+            self,
+            lambda assets: AssetsDefinition.with_resources(assets, resource_defs),
+        )
+
+    def with_prefix_or_group(
+        self,
+        output_asset_key_replacements: Optional[Mapping[AssetKey, AssetKey]] = None,
+        input_asset_key_replacements: Optional[Mapping[AssetKey, AssetKey]] = None,
+        group_names_by_key: Optional[Mapping[AssetKey, str]] = None,
+    ) -> "CacheableAssetsDefinition":
+        return WrappedCacheableAssetsDefinition(
+            self._unique_id,
+            self,
+            lambda assets: AssetsDefinition.with_prefix_or_group(
+                assets,
+                output_asset_key_replacements=output_asset_key_replacements,
+                input_asset_key_replacements=input_asset_key_replacements,
+                group_names_by_key=group_names_by_key,
+            ),
+        )
+
+
+class WrappedCacheableAssetsDefinition(CacheableAssetsDefinition):
+    """
+    Wraps an instance of CacheableAssetsDefinition and applies a function to the
+    generated AssetsDefinition objects. This lets e.g. users define resources on
+    the cacheable assets at repo creation time which are not actually bound until
+    the assets themselves are created.
+    """
+
+    def __init__(
+        self,
+        unique_id: str,
+        wrapped: CacheableAssetsDefinition,
+        fn: Callable[[AssetsDefinition], AssetsDefinition],
+    ):
+        super().__init__(unique_id)
```

Comment:
This unique_id doesn't need to be passed in separately because it should always be a function of wrapped.unique_id (currently, we're just copying it in place but we could consider hashing in a constant value or something)